hibernate-core/src/main/java/org/hibernate/event/internal/DefaultRefreshEventListener.java

```diff
@@ -168,9 +172,30 @@ public void onRefresh(RefreshEvent event, Map refreshedAlready) {
 
 		String previousFetchProfile = source.getLoadQueryInfluencers().getInternalFetchProfile();
 		source.getLoadQueryInfluencers().setInternalFetchProfile( "refresh" );
-		Object result = persister.load( id, object, event.getLockOptions(), source );
+
+		// Use the entity's current LockMode if it is greater than event.getLockMode(),
+		// unless the current LockMode is WRITE (special case described below).
+		final LockMode currentLockMode = e == null ? null : e.getLockMode();
+		LockOptions lockOptionsToUse = event.getLockOptions();
+		boolean resetLockModeToWrite = false;
```

Comment:
My last review comment should have been attached here, but as I replied to our previous conversations (line comments on out-of-date lines) it is not showing as part of the review...

I think this can be made much simpler. Working on an alternate PR for just this last little piece (WRITE)

Reply:
Oops, sorry about that. Here's your comment:

"If the LockMode is WRITE, we already own the exclusive lock on that row, its impossible for another process to update it meaning that there ought to be no changes to "pull back". From this perspective we could just skip the refresh.

However, the one use-case we do need to worry about (at least the only one I can think of) are db generated values where the user is not taking advantage of Hibernate's generated-value support. The JPA-standard way to deal with that is to refresh after the persist/merge.

So I think the general case for currentLockMode == WRITE could be to fallback to READ for the load, but to reset the entry's LockMode to WRITE after.

BTW, wouldn't we have the same problem for PESSIMISTIC_WRITE (and PESSIMISTIC_READ possibly)?"